planner/cascades/transformation_rules.go

```diff
@@ -229,6 +235,88 @@ func (r *EnumeratePaths) OnTransform(old *memo.ExprIter) (newExprs []*memo.Group
 	return newExprs, true, false, nil
 }
 
+// PushAggDownGather splits Aggregation to two stages, final and partial1,
+// and pushed the partial Aggregation down to the child of TableGather.
+type PushAggDownGather struct {
+}
+
+// GetPattern implements Transformation interface. The pattern of this rule
+// is `Aggregation -> TableGather`.
+func (r *PushAggDownGather) GetPattern() *memo.Pattern {
+	return memo.BuildPattern(
+		memo.OperandAggregation,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandTableGather, memo.EngineTiDBOnly),
+	)
+}
+
+// Match implements Transformation interface.
+func (r *PushAggDownGather) Match(expr *memo.ExprIter) bool {
+	agg := expr.GetExpr().ExprNode.(*plannercore.LogicalAggregation)
+	for _, aggFunc := range agg.AggFuncs {
+		if aggFunc.Mode != aggregation.CompleteMode {
```

Comment:
Need we check this?

Reply:
If we don't check this, the exploration won't stop. For example: we have `Agg(complete) -> Gather` at first. After the first time we explore the group, we will have `[Agg(complete, explored), Agg(final, unexplored)]`, the finalAgg still can apply this rule.